File: packages/sdk-python/src/primust/pipeline.py
"""
Primust Python SDK — Pipeline class.

Privacy invariant: raw content NEVER leaves the customer environment.
Only commitment hashes (poseidon2/sha256) transit to the Primust API.

Usage (new Run-based API):
    from primust import Pipeline

    p = Pipeline(api_key="pk_live_org001_us_secret", workflow_id="wf_onboard")
    run = p.open()
    result = run.record(check="pii_check", manifest_id="...",
                        input=my_data, check_result="pass")
    vpec = run.close()

Usage (legacy session-based API):
    p = Pipeline(api_key="pk_live_org001_us_secret", workflow_id="wf_onboard")
    session = p.open_check("pii_check", "manifest_001")
    result = p.record(session, input=my_data, check_result="pass")
    vpec = p.close()
"""

from __future__ import annotations

import datetime
import os
import uuid
from dataclasses import dataclass, field
from datetime import timezone
from pathlib import Path
from typing import Any, Optional

import httpx

from primust_artifact_core import (
    ZK_IS_BLOCKING,
    commit,
    commit_output,
)

from .models import (
    CheckResult,
    GovernanceGap,
    ManifestRegistration,
    ProofLevel,
    ProofLevelBreakdown,
    RecordResult,
    VPEC,
    VisibilityMode,
)
from .models import CheckSession as _ModelsCheckSession
from .models import ReviewSession as _ModelsReviewSession
from .queue import LocalQueue
from .run import Run
from .transport import PrimustTransport

# Re-export for SDK consumers
__all__ = [
    "Pipeline",
    "CheckSession",
    "ReviewSession",
    "RecordResult",
    "ResumedContext",
    "ZK_IS_BLOCKING",
]

BASE_URL = "https://api.primust.com"


# ── Legacy session types (kept for backwards compatibility) ──


@dataclass
class CheckSession:
    """Returned by Pipeline.open_check()."""

    check_name: str
    manifest_id: str
    manifest_hash: str | None
    check_open_tst: str | None  # RFC 3161 timestamp (stub: ISO string)


@dataclass
class ReviewSession(CheckSession):
    """Returned by Pipeline.open_review() — for witnessed-level records."""

    reviewer_key_id: str = ""
    min_duration_seconds: int = 1800
    opened_at: str = ""


@dataclass
class ResumedContext:
    """Returned by Pipeline.resume_from_lineage()."""

    run_id: str
    surface_id: str
    delegation_context: dict[str, Any] = field(default_factory=dict)


class Pipeline:
    """
    Primust governance pipeline.

    Two usage modes:
    1. Run-based (recommended): pipeline.open() → Run → run.record() → run.close()
    2. Legacy session-based: pipeline.open_check() → pipeline.record() → pipeline.close()

    All commitment hashing happens locally. Raw content NEVER transits.
    """

    def __init__(
        self,
        api_key: Optional[str] = None,
        workflow_id: str = "default",
        *,
        surface_id: Optional[str] = None,
        policy: list[str] | str | None = None,
        process_context_hash: str | None = None,
        base_url: str = BASE_URL,
        http_client: httpx.Client | None = None,
        queue_path: Optional[Path] = None,
        _base_url: Optional[str] = None,  # alias for testing
    ) -> None:
        resolved_key = api_key or os.environ.get("PRIMUST_API_KEY")
        if not resolved_key:
            raise ValueError(
                "api_key is required. Pass it directly or set PRIMUST_API_KEY env var."
            )

        self.api_key = resolved_key
        self.workflow_id = workflow_id
        self.surface_id = surface_id
        self.policy = policy
        self.process_context_hash = process_context_hash
        self.test_mode = resolved_key.startswith("pk_test_")
        self.base_url = (_base_url or base_url).rstrip("/")

        # Transport + queue for Run-based API
        self._queue = LocalQueue(queue_path) if queue_path else LocalQueue()
        api_base = self.base_url
        if not api_base.endswith("/api/v1"):
            api_base = f"{api_base}/api/v1"
        self._transport = PrimustTransport(
            api_key=resolved_key,
            queue=self._queue,
            base_url=api_base,
        )

        # Legacy httpx client for session-based API
        self._client = http_client or httpx.Client(
            base_url=self.base_url,
            headers={"X-API-Key": resolved_key},
            timeout=30.0,
        )
        self._run_id: str | None = None
        self._surface_id: str | None = surface_id
        self._manifest_hashes: dict[str, str] = {}
        self._prior_manifest_hashes: dict[str, str] = {}
        self._closed = False
        self._org_id: Optional[str] = None
        self._policy_snapshot_hash: Optional[str] = None

    # ══════════════════════════════════════════════════════════════════
    # Run-based API (recommended)
    # ══════════════════════════════════════════════════════════════════

    def open(self, policy_pack_id: Optional[str] = None) -> Run:
        """
        Open a new governed process run.
        Returns a Run. Call run.record() for each governance check.
        Close with run.close() to issue the VPEC.
        """
        run_id = f"run_{uuid.uuid4().hex}"
        opened_at = datetime.datetime.utcnow().isoformat() + "Z"

        open_payload: dict[str, Any] = {
            "run_id": run_id,
            "workflow_id": self.workflow_id,
            "environment": "test" if self.test_mode else "production",
            "opened_at": opened_at,
        }
        if policy_pack_id:
            open_payload["policy_pack_id"] = policy_pack_id
        if self.surface_id:
            open_payload["surface_id"] = self.surface_id

        response = self._transport.post_open_run(open_payload)

        # Use server-assigned run_id (server generates its own, ignores client's)
        server_run_id = response.get("run_id", run_id)
        self._org_id = self._org_id or response.get("org_id", "unknown")
        self._policy_snapshot_hash = response.get("policy_snapshot_hash", "")

        return Run(
            run_id=server_run_id,
            workflow_id=self.workflow_id,
            org_id=self._org_id,
            policy_snapshot_hash=self._policy_snapshot_hash or "",
            transport=self._transport,
            test_mode=self.test_mode,
        )

    def register_check(self, manifest: dict) -> ManifestRegistration:
        """
        Register a check manifest. Call once per manifest version.
        Returns ManifestRegistration with manifest_id.
        """
        response = self._transport.post_manifest(manifest)
        return ManifestRegistration(
            manifest_id=response["manifest_id"],
            name=manifest.get("name", ""),
            registered_at=response.get("registered_at", ""),
        )

    def pending_queue_count(self) -> int:
        """Number of records/closes waiting in the local queue."""
        return self._queue.count()

    def flush_queue(self) -> int:
        """Attempt to flush queued records. Returns items flushed."""
        return self._transport.flush_queue()

    # ══════════════════════════════════════════════════════════════════
    # Legacy session-based API
    # ══════════════════════════════════════════════════════════════════

    def _ensure_run(self, surface_id: str = "default") -> str:
        """Lazily open a run on first check."""
        if self._run_id:
            return self._run_id

        body: dict[str, Any] = {
            "workflow_id": self.workflow_id,
            "surface_id": surface_id,
            "policy_pack_id": self.policy if isinstance(self.policy, str) else "default",
        }
        if self.process_context_hash:
            body["process_context_hash"] = self.process_context_hash

        resp = self._client.post("/api/v1/runs", json=body)
        resp.raise_for_status()
        data = resp.json()
        self._run_id = data["run_id"]
        return self._run_id

    def open_check(
        self,
        check: str,
        manifest_id: str,
        **kwargs: Any,
    ) -> CheckSession:
        """Open a check session. Fetches check_open_tst immediately."""
        self._ensure_run(kwargs.get("surface_id", "default"))
        now = datetime.datetime.now(timezone.utc).isoformat()

        return CheckSession(
            check_name=check,
            manifest_id=manifest_id,
            manifest_hash=self._manifest_hashes.get(manifest_id, manifest_id),
            check_open_tst=now,
        )

    def open_review(
        self,
        check: str,
        manifest_id: str,
        reviewer_key_id: str,
        min_duration_seconds: int = 1800,
        **kwargs: Any,
    ) -> ReviewSession:
        """Open a review session for witnessed-level records."""
        self._ensure_run(kwargs.get("surface_id", "default"))
        now = datetime.datetime.now(timezone.utc).isoformat()

        return ReviewSession(
            check_name=check,
            manifest_id=manifest_id,
            manifest_hash=self._manifest_hashes.get(manifest_id, manifest_id),
            check_open_tst=now,
            reviewer_key_id=reviewer_key_id,
            min_duration_seconds=min_duration_seconds,
            opened_at=now,
        )

    def record(
        self,
        check_session: CheckSession,
        input: Any,  # noqa: A002 — matches spec name
        check_result: str,
        *,
        output: Any = None,
        reviewer_signature: str | None = None,
        display_content: Any = None,
        rationale: str | None = None,
        skip_rationale: str | None = None,
        **kwargs: Any,
    ) -> RecordResult:
        """
        Record a check execution (legacy session-based API).
        Raw content is NEVER sent — only commitment hashes.
        """
        assert self._run_id, "Pipeline not opened"

        now = datetime.datetime.now(timezone.utc).isoformat()

        # Enforce min_duration_seconds for review sessions
        if isinstance(check_session, ReviewSession) and check_session.opened_at:
            opened_dt = datetime.datetime.fromisoformat(check_session.opened_at)
            now_dt = datetime.datetime.fromisoformat(now)
            elapsed = (now_dt - opened_dt).total_seconds()
            if elapsed < check_session.min_duration_seconds:
                raise ValueError(
                    f"Review duration {elapsed:.1f}s is below minimum "
                    f"{check_session.min_duration_seconds}s (check_timing_suspect)"
                )

        # Compute commitment hashes locally
        input_bytes = _to_bytes(input)
        commitment_hash, algorithm = commit(input_bytes)

        output_commitment_str: str | None = None
        if output is not None:
            output_bytes = _to_bytes(output)
            output_commitment_str, _ = commit_output(output_bytes)

        skip_rationale_hash: str | None = None
        if skip_rationale is not None:
            skip_hash, _ = commit(skip_rationale.encode("utf-8"))
            skip_rationale_hash = skip_hash

        # Build reviewer_credential (for witnessed records)
        reviewer_credential: dict[str, Any] | None = None
        if reviewer_signature is not None and isinstance(check_session, ReviewSession):
            display_hash: str | None = None
            rationale_hash: str | None = None

            if display_content is not None:
                dh, _ = commit(_to_bytes(display_content))
                display_hash = dh
            if rationale is not None:
                rh, _ = commit(rationale.encode("utf-8"))
                rationale_hash = rh

            reviewer_credential = {
                "reviewer_key_id": check_session.reviewer_key_id,
                "key_binding": "org_managed",
                "role": "reviewer",
                "org_credential_ref": None,
                "reviewer_signature": reviewer_signature,
                "display_hash": display_hash or "",
                "rationale_hash": rationale_hash or "",
                "signed_content_hash": commitment_hash,
                "open_tst": check_session.check_open_tst or "",
                "close_tst": now,
            }

        proof_level = "execution"
        if reviewer_credential:
            proof_level = "witnessed"

        body: dict[str, Any] = {
            "manifest_id": check_session.manifest_id,
            "commitment_hash": commitment_hash,
            "commitment_algorithm": algorithm,
            "commitment_type": "input_only" if output is None else "input_output",
            "check_result": check_result,
            "proof_level_achieved": proof_level,
            "check_open_tst": check_session.check_open_tst,
            "check_close_tst": now,
            "idempotency_key": f"idem_{uuid.uuid4().hex[:16]}",
        }

        if output_commitment_str:
            body["output_commitment"] = output_commitment_str
        if skip_rationale_hash:
            body["skip_rationale_hash"] = skip_rationale_hash
        if reviewer_credential:
            body["reviewer_credential"] = reviewer_credential

        resp = self._client.post(
            f"/api/v1/runs/{self._run_id}/records",
            json=body,
        )
        resp.raise_for_status()
        data = resp.json()

        return RecordResult(
            record_id=data["record_id"],
            commitment_hash=commitment_hash,
            output_commitment=output_commitment_str,
            commitment_algorithm=algorithm,
            proof_level=proof_level,
            recorded_at=now,
            chain_hash=data["chain_hash"],
        )

    def record_delegation(self, outbound_context: dict[str, Any]) -> dict[str, Any]:
        """Emit external_boundary_traversal gap (Informational). Return lineage token."""
        assert self._run_id, "Pipeline not opened"
        return {
            "token": f"lt_{uuid.uuid4().hex[:16]}",
            "run_id": self._run_id,
            "surface_id": self._surface_id or "default",
            "delegation_context": outbound_context,
            "issued_at": datetime.datetime.now(timezone.utc).isoformat(),
        }

    def resume_from_lineage(self, token: dict[str, Any]) -> ResumedContext:
        """Validate lineage token. On failure: lineage_token_missing gap (High)."""
        return ResumedContext(
            run_id=token.get("run_id", ""),
            surface_id=token.get("surface_id", ""),
            delegation_context=token.get("delegation_context", {}),
        )

    def close(
        self, *, partial: bool = False, request_zk: bool = False
    ) -> dict[str, Any]:
        """Close the run and return signed VPEC (legacy API returns raw dict)."""
        assert self._run_id, "Pipeline not opened"
        assert not self._closed, "Pipeline already closed"

        resp = self._client.post(
            f"/api/v1/runs/{self._run_id}/close",
            json={"partial": partial, "request_zk": request_zk},
        )
        resp.raise_for_status()
        self._closed = True
        return resp.json()


def _to_bytes(value: Any) -> bytes:
    """Convert a value to bytes for commitment hashing."""
    if isinstance(value, bytes):
        return value
    if isinstance(value, str):
        return value.encode("utf-8")
    import json
    return json.dumps(value, sort_keys=True, separators=(",", ":")).encode("utf-8")

File: packages/sdk-js/src/pipeline.ts
/**
 * Primust TypeScript SDK — Pipeline class.
 *
 * Privacy invariant: raw content NEVER leaves the customer environment.
 * Only commitment hashes (poseidon2/sha256) transit to the Primust API.
 */

import {
  commit,
  commitOutput,
  ZK_IS_BLOCKING,
  canonical,
} from '@primust/artifact-core';
import type { CommitmentResult } from '@primust/artifact-core';

import { Run } from './run.js';

// Re-export constant
export { ZK_IS_BLOCKING };

// ── Types ──

export interface PipelineConfig {
  apiKey: string; // pk_live_xxx | pk_sb_xxx
  workflowId: string;
  policy?: string[] | string;
  processContextHash?: string;
  baseUrl?: string;
  fetch?: typeof globalThis.fetch;
}

export interface CheckSession {
  checkName: string;
  manifestId: string;
  manifestHash: string | null;
  checkOpenTst: string | null;
}

export interface ReviewSession extends CheckSession {
  reviewerKeyId: string;
  minDurationSeconds: number;
  openedAt: string;
}

export interface RecordResult {
  recordId: string;
  chainHash: string;
  commitmentHash: string;
  outputCommitment: string | null;
}

export interface RecordOptions {
  output?: unknown;
  reviewerSignature?: string;
  displayContent?: unknown;
  rationale?: string;
  skipRationale?: string;
}

export interface CloseOptions {
  partial?: boolean;
  requestZk?: boolean;
}

export interface ResumedContext {
  runId: string;
  surfaceId: string;
  delegationContext: Record<string, unknown>;
}

export interface PrimustLogEvent {
  primust_record_id: string;
  primust_commitment_hash: string;   // poseidon2:hex — linkage anchor
  primust_check_result: string;
  primust_proof_level: string;
  primust_workflow_id: string;
  primust_run_id: string;
  primust_recorded_at: string;       // ISO 8601
  gap_types_emitted?: string[];      // only if options.includeGapTypes
}

export interface LoggerOptions {
  includeGapTypes?: boolean;         // default false
}

// ── Helpers ──

function toBytes(value: unknown): Uint8Array {
  if (value instanceof Uint8Array) return value;
  if (typeof value === 'string') return new TextEncoder().encode(value);
  return new TextEncoder().encode(canonical(value));
}

// ── Pipeline ──

export class Pipeline {
  private readonly apiKey: string;
  private readonly workflowId: string;
  private readonly policy: string[] | string | undefined;
  private readonly processContextHash: string | undefined;
  private readonly baseUrl: string;
  private readonly _fetch: typeof globalThis.fetch;
  private runId: string | null = null;
  private surfaceId: string | null = null;
  private closed = false;

  /** @internal For config drift detection */
  _priorManifestHashes: Record<string, string> = {};
  /** @internal Current manifest hashes */
  _manifestHashes: Record<string, string> = {};

  private _loggerCallback: ((event: PrimustLogEvent) => void) | null = null;
  private _loggerOptions: LoggerOptions = {};

  constructor(config: PipelineConfig) {
    this.apiKey = config.apiKey;
    this.workflowId = config.workflowId;
    this.policy = config.policy;
    this.processContextHash = config.processContextHash;
    this.baseUrl = (config.baseUrl ?? 'https://api.primust.com').replace(/\/+$/, '');
    this._fetch = config.fetch ?? globalThis.fetch.bind(globalThis);
  }

  /**
   * Register a log callback for SIEM linkage.
   *
   * The callback receives a PrimustLogEvent on every p.record() call.
   * Write primust_commitment_hash to your existing logging infrastructure
   * (Splunk, Datadog, CloudWatch, etc.) to create an auditable linkage
   * between your application logs and the Primust VPEC.
   *
   * Auditor verification:
   *   1. Search SIEM: WHERE primust_commitment_hash = <value>
   *   2. primust-verify hash <plaintext_input> → confirms hash matches
   *   3. Primust VPEC proves chain integrity and timestamp independence
   *
   * Raw content is never passed to this callback.
   */
  setLogger(
    callback: (event: PrimustLogEvent) => void,
    options?: LoggerOptions,
  ): void {
    this._loggerCallback = callback;
    if (options) this._loggerOptions = options;
  }

  private _invokeLogger(event: PrimustLogEvent): void {
    if (!this._loggerCallback) return;
    try {
      this._loggerCallback(event);
    } catch {
      // Exceptions in callback are caught — never propagate.
    }
  }

  private async api(
    method: string,
    path: string,
    body?: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    const resp = await this._fetch(`${this.baseUrl}${path}`, {
      method,
      headers: {
        'X-API-Key': this.apiKey,
        'Content-Type': 'application/json',
      },
      body: body ? JSON.stringify(body) : undefined,
    });
    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`API ${resp.status}: ${text}`);
    }
    return resp.json() as Promise<Record<string, unknown>>;
  }

  private async ensureRun(surfaceId = 'default'): Promise<string> {
    if (this.runId) return this.runId;

    const body: Record<string, unknown> = {
      workflow_id: this.workflowId,
      surface_id: surfaceId,
      policy_pack_id: typeof this.policy === 'string' ? this.policy : 'default',
    };
    if (this.processContextHash) {
      body.process_context_hash = this.processContextHash;
    }

    const data = await this.api('POST', '/api/v1/runs', body);
    this.runId = data.run_id as string;
    return this.runId;
  }

  /**
   * Open a new governed process run.
   * Returns a Run. Call run.record() for each governance check.
   * Close with run.close() to issue the VPEC.
   */
  async open(policyPackId?: string): Promise<Run> {
    const body: Record<string, unknown> = {
      workflow_id: this.workflowId,
      environment: this.apiKey.startsWith('pk_test_') || this.apiKey.startsWith('pk_sb_') ? 'test' : 'production',
      opened_at: new Date().toISOString(),
    };
    if (policyPackId) {
      body.policy_pack_id = policyPackId;
    }

    const data = await this.api('POST', '/api/v1/runs', body);
    const serverRunId = (data.run_id as string) ?? `run_${crypto.randomUUID().replace(/-/g, '')}`;
    const orgId = (data.org_id as string) ?? 'unknown';
    const policySnapshotHash = (data.policy_snapshot_hash as string) ?? '';
    const testMode = this.apiKey.startsWith('pk_test_') || this.apiKey.startsWith('pk_sb_');

    return new Run({
      runId: serverRunId,
      workflowId: this.workflowId,
      orgId,
      policySnapshotHash,
      fetch: this._fetch,
      baseUrl: this.baseUrl,
      apiKey: this.apiKey,
      testMode,
      loggerCallback: this._loggerCallback,
      loggerOptions: this._loggerOptions,
    });
  }

  async openCheck(
    check: string,
    manifestId: string,
    _options?: Record<string, unknown>,
  ): Promise<CheckSession> {
    await this.ensureRun();
    const now = new Date().toISOString();

    // Populate _manifestHashes: compute hash from manifest data if provided,
    // otherwise use manifestId as the hash identity.
    if (_options && Object.keys(_options).length > 0) {
      const manifestBytes = new TextEncoder().encode(canonical(_options));
      const { hash } = commit(manifestBytes);
      this._manifestHashes[manifestId] = hash;
    } else if (!this._manifestHashes[manifestId]) {
      const { hash } = commit(new TextEncoder().encode(manifestId));
      this._manifestHashes[manifestId] = hash;
    }

    return {
      checkName: check,
      manifestId,
      manifestHash: this._manifestHashes[manifestId] ?? manifestId,
      checkOpenTst: now,
    };
  }

  async openReview(
    check: string,
    manifestId: string,
    options: { reviewerKeyId: string; minDurationSeconds?: number },
  ): Promise<ReviewSession> {
    await this.ensureRun();
    const now = new Date().toISOString();

    // Populate _manifestHashes if not already present
    if (!this._manifestHashes[manifestId]) {
      const { hash } = commit(new TextEncoder().encode(manifestId));
      this._manifestHashes[manifestId] = hash;
    }

    return {
      checkName: check,
      manifestId,
      manifestHash: this._manifestHashes[manifestId] ?? manifestId,
      checkOpenTst: now,
      reviewerKeyId: options.reviewerKeyId,
      minDurationSeconds: options.minDurationSeconds ?? 1800,
      openedAt: now,
    };
  }

  async record(
    checkSession: CheckSession,
    input: unknown,
    checkResult: string,
    options: RecordOptions = {},
  ): Promise<RecordResult> {
    if (!this.runId) throw new Error('Pipeline not opened');

    const now = new Date().toISOString();

    // Compute commitment hashes locally — raw content NEVER sent
    const inputBytes = toBytes(input);
    const { hash: commitmentHash, algorithm: commitmentAlgorithm } = commit(inputBytes);

    let outputCommitment: string | null = null;
    if (options.output !== undefined) {
      const outputBytes = toBytes(options.output);
      const result = commitOutput(outputBytes);
      outputCommitment = result.hash;
    }

    let skipRationaleHash: string | null = null;
    if (options.skipRationale !== undefined) {
      const { hash } = commit(new TextEncoder().encode(options.skipRationale));
      skipRationaleHash = hash;
    }

    // Enforce min_duration_seconds for review sessions at record() time
    if ('reviewerKeyId' in checkSession) {
      const rs = checkSession as ReviewSession;
      const openedMs = new Date(rs.openedAt).getTime();
      const nowMs = new Date(now).getTime();
      const elapsedSec = (nowMs - openedMs) / 1000;
      if (elapsedSec < rs.minDurationSeconds) {
        throw new Error(
          `Review duration ${elapsedSec.toFixed(1)}s is below minimum ${rs.minDurationSeconds}s (check_timing_suspect)`,
        );
      }
    }

    // Build reviewer_credential for witnessed records
    let reviewerCredential: Record<string, unknown> | undefined;
    if (options.reviewerSignature && 'reviewerKeyId' in checkSession) {
      const rs = checkSession as ReviewSession;
      let displayHash = '';
      let rationaleHash = '';

      if (options.displayContent !== undefined) {
        const { hash } = commit(toBytes(options.displayContent));
        displayHash = hash;
      }
      if (options.rationale !== undefined) {
        const { hash } = commit(new TextEncoder().encode(options.rationale));
        rationaleHash = hash;
      }

      reviewerCredential = {
        reviewer_key_id: rs.reviewerKeyId,
        key_binding: 'software',
        role: 'reviewer',
        org_credential_ref: null,
        reviewer_signature: options.reviewerSignature,
        display_hash: displayHash,
        rationale_hash: rationaleHash,
        signed_content_hash: commitmentHash,
        open_tst: checkSession.checkOpenTst ?? '',
        close_tst: now,
      };
    }

    const proofLevel = reviewerCredential ? 'witnessed' : 'execution';

    const body: Record<string, unknown> = {
      manifest_id: checkSession.manifestId,
      commitment_hash: commitmentHash,
      commitment_algorithm: commitmentAlgorithm,
      commitment_type: 'input_commitment',
      check_result: checkResult,
      proof_level_achieved: proofLevel,
      check_open_tst: checkSession.checkOpenTst,
      check_close_tst: now,
      idempotency_key: `idem_${crypto.randomUUID().replace(/-/g, '').slice(0, 16)}`,
    };

    if (outputCommitment) body.output_commitment = outputCommitment;
    if (skipRationaleHash) body.skip_rationale_hash = skipRationaleHash;
    if (reviewerCredential) body.reviewer_credential = reviewerCredential;

    // Logger callback — called after commitment_hash computed, before API call.
    this._invokeLogger({
      primust_record_id: body.idempotency_key as string,
      primust_commitment_hash: commitmentHash,
      primust_check_result: checkResult,
      primust_proof_level: proofLevel,
      primust_workflow_id: this.workflowId,
      primust_run_id: this.runId!,
      primust_recorded_at: now,
      gap_types_emitted: [],
    });

    const data = await this.api('POST', `/api/v1/runs/${this.runId}/records`, body);

    return {
      recordId: data.record_id as string,
      chainHash: data.chain_hash as string,
      commitmentHash,
      outputCommitment,
    };
  }

  async recordDelegation(
    context: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    if (!this.runId) throw new Error('Pipeline not opened');

    return {
      token: `lt_${crypto.randomUUID().replace(/-/g, '').slice(0, 16)}`,
      run_id: this.runId,
      surface_id: this.surfaceId ?? 'default',
      delegation_context: context,
      issued_at: new Date().toISOString(),
    };
  }

  async resumeFromLineage(
    token: Record<string, unknown>,
  ): Promise<ResumedContext> {
    return {
      runId: (token.run_id as string) ?? '',
      surfaceId: (token.surface_id as string) ?? '',
      delegationContext: (token.delegation_context as Record<string, unknown>) ?? {},
    };
  }

  async close(options: CloseOptions = {}): Promise<Record<string, unknown>> {
    if (!this.runId) throw new Error('Pipeline not opened');
    if (this.closed) throw new Error('Pipeline already closed');

    const data = await this.api('POST', `/api/v1/runs/${this.runId}/close`, {
      partial: options.partial ?? false,
      request_zk: options.requestZk ?? false,
    });

    this.closed = true;
    return data;
  }
}
